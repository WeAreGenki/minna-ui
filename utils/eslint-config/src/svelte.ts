/**
 * Minna UI Svelte & Sapper ESLint config add-on.
 * @fileoverview Provides extra config for Svelte & Sapper projects. This
 * config is and add-on which should extend the base config preset.
 * @see https://eslint.org/docs/user-guide/configuring
 */

/* eslint-disable sort-keys */

interface BlockAttributes {
  type: string;
}

export = {
  plugins: ['svelte3'],
  settings: {
    'svelte3/ignore-styles': (attr: BlockAttributes) =>
      attr.type === 'text/postcss',
  },
  overrides: [
    // Svelte components
    {
      files: ['*.svelte'],
      rules: {
        // import order cannot be determined correctly in .svelte components
        'import/first': 'off',
        // components are compiled at build-time so devDependencies are OK
        'import/no-extraneous-dependencies': [
          'error',
          { devDependencies: true },
        ],
        // Svelte abuses the syntax with `export let`
        'import/no-mutable-exports': 'off',
      },
    },
  ],
};
