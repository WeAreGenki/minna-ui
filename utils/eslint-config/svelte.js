/**
 * Minna UI Svelte & Sapper ESLint config add-on.
 *
 * @fileoverview Provides extra config for Svelte & Sapper projects. This
 * config is and add-on which should extend the base config preset.
 * @see https://eslint.org/docs/user-guide/configuring
 */

/* eslint-disable sort-keys */

'use strict';

/** @type {import('./types').ESLintConfig} */
module.exports = {
  plugins: ['svelte3'],
  settings: {
    /**
     * @param {import('./types').BlockAttributes} attr - Tag attributes.
     * @returns {boolean} If style block should be ignored.
     */
    'svelte3/ignore-styles': (attr) => attr.type === 'text/postcss',
    // TODO: Uncomment once false positives are handled better
    // 'svelte3/lint-template': true,
  },
  overrides: [
    // Svelte components
    {
      files: ['*.svelte'],
      rules: {
        // import order cannot be determined correctly in .svelte components
        'import/first': 'off',
        // components are compiled at build-time so devDependencies are OK
        'import/no-extraneous-dependencies': [
          'error',
          { devDependencies: true },
        ],
        // Svelte abuses the syntax with `export let`
        'import/no-mutable-exports': 'off',
      },
    },
  ],
};
