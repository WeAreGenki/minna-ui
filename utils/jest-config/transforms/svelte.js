/**
 * Jest transform for compiling Svelte components into JavaScript.
 *
 * NOTE: `svelte` is not listed as a dependency so developers can use this
 * jest preset for non-svelte projects.
 */

'use strict';

const { compile } = require('svelte/compiler'); // eslint-disable-line import/no-extraneous-dependencies

/** @type {import('../types').JestTransformer} */
exports.process = (src, filename) => {
  // Strip out <style> tags to prevent errors when unable to parse PostCSS etc.
  const re = /<style[^>]*>[\S\s]*?<\/style>/g;
  const normalised = src.replace(re, '');

  const result = compile(normalised, {
    accessors: true, // Easy access to component props for tests
    css: false,
    dev: true,
    filename,
    format: 'cjs',
    preserveWhitespace: true,
  });

  const esInterop =
    'Object.defineProperty(exports, "__esModule", { value: true });';

  return {
    code: result.js.code + esInterop,
    map: result.js.map,
  };
};
