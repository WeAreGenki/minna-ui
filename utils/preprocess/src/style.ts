/* eslint-disable no-restricted-syntax */

import merge from 'deepmerge';
import postcss from 'postcss';
import postcssLoadConfig from 'postcss-load-config';
import syntax from 'postcss-scss';
// FIXME: Replace once svelte is fixed
// import { Preprocessor } from 'svelte/types/preprocess';
import { Preprocessor } from './types';

/**
 * Minna UI svelte style preprocessor.
 *
 * @param opts - PostCSS options.
 */
export const style = (
  opts: postcss.ProcessOptions = {},
): Preprocessor => async ({ attributes, content, filename }) => {
  if (attributes.type !== 'text/postcss') return;

  // merge user provided options into default context
  const context = merge(
    {
      from: filename,
      map: {
        annotation: false,
        inline: false,
      },
      syntax,
      to: filename,
    },
    opts,
  );

  /**
   * Compile PostCSS code into CSS.
   */
  try {
    const { plugins, options } = await postcssLoadConfig(context);
    const result = await postcss(plugins).process(content, options);

    for (const err of result.warnings()) {
      // eslint-disable-next-line no-console
      console.warn(err.toString());
    }

    const dependencies: string[] = [];

    // register dependencies so rollup can watch them for changes
    for (const msg of result.messages) {
      if (msg.type === 'dependency') {
        dependencies.push(msg.file);
      }
    }

    // eslint-disable-next-line consistent-return
    return {
      code: result.css,
      dependencies,
      map: result.map,
    };
  } catch (err) {
    if (err.name === 'CssSyntaxError') {
      process.stderr.write(err.message + err.showSourceCode());
    } else {
      // eslint-disable-next-line no-console
      throw console.error(err);
    }
  }
};
