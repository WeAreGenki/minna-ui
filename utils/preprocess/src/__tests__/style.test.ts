/** @jest-environment node */

// TODO: Add test for returned `dependencies`

import fs from 'fs';
import { promisify } from 'util';
import { compile, preprocess } from 'svelte/compiler';
import nested from 'postcss-nested';
import { style } from '../style';

// eslint-disable-next-line security/detect-non-literal-fs-filename
const readFile = promisify(fs.readFile);

// don't require() component to avoid Jest transform
const componentPath = require.resolve('@minna-ui/jest-config/fixtures/TestComponent.svelte');

const preprocessOpts = {
  style: style({
    /* stylelint-disable-next-line */ // false positive! ðŸ‘¾
    plugins: [nested],
  }),
};

const sourceSimple = `
<div id="target">test</div>

<style type="text/postcss">
  #target {
    color: red;

    &:focus {
      background: #000;
    }
  }
</style>
`;
const sourceNoTypeAttr = `
<div id="target">test</div>

<style>
  #target {
    color: red;

    &:focus {
      background: #000;
    }
  }
</style>
`;
const sourceBadSyntax = `
<style type="text/postcss">
  #target {
    color: red;
</style>
`;
let source = '';

beforeAll(async () => {
  source = await readFile(componentPath, 'utf8');
});

describe('Svelte style preprocessor', () => {
  it('processes a simple component', async () => {
    expect.assertions(3);
    const output = preprocess(sourceSimple, preprocessOpts);
    await expect(output).resolves.toBeDefined();
    const result = (await output).toString();
    expect(result).not.toMatch('&:focus');
    expect(result).toMatchSnapshot();
  });

  it('does not process without type attribute', async () => {
    expect.assertions(2);
    const result = (await preprocess(sourceNoTypeAttr, preprocessOpts)).toString();
    expect(result).toMatch('&:focus');
    expect(result).toMatchSnapshot();
  });

  it('processes a component', async () => {
    expect.assertions(4);
    const output = preprocess(source, preprocessOpts);
    await expect(output).resolves.toBeDefined();
    const result = (await output).toString();
    expect(result).not.toBe(false);
    expect(result).not.toEqual('');
    expect(result).toMatchSnapshot();
  });

  it("compiles a component's CSS", async () => {
    expect.assertions(2);
    const processed = await preprocess(source, preprocessOpts);
    const result = compile(processed.toString());
    expect(result.css.code).not.toBe(false);
    expect(result.css.code).toMatchSnapshot();
  });

  it('prints error on bad CSS syntax', async () => {
    expect.assertions(1);
    const spy = jest.spyOn(process.stderr, 'write');
    spy.mockImplementation(() => true);
    await preprocess(sourceBadSyntax, preprocessOpts);
    expect(spy).toHaveBeenCalledWith(expect.stringContaining('<css input>:2:3: Unclosed block'));
    spy.mockRestore();
  });
});
