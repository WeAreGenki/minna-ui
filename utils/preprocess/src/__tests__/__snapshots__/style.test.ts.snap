// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Svelte style preprocessor adds a banner comment 1`] = `
"
<div id=\\"target\\">test</div>

<style type=\\"text/postcss\\">
  #target {
    color: red;
  }

    #target:focus {
      background: #000;
    }
</style>
"
`;

exports[`Svelte style preprocessor compiles a component's CSS 1`] = `
"@media screen and (min-width: 30em){body{margin:0
  }}.text.svelte-1ghwuak{color:rgb(255, 0, 0)}.text.svelte-1ghwuak:focus{background:rgb(0, 0, 0)}.wrapper.svelte-1ghwuak>.text.svelte-1ghwuak{margin-top:1rem}"
`;

exports[`Svelte style preprocessor does not process without type attribute 1`] = `
"
<div id=\\"target\\">test</div>

<style>
  #target {
    color: red;

    &:focus {
      background: #000;
    }
  }
</style>
"
`;

exports[`Svelte style preprocessor processes a component 1`] = `
"<!--
  TEST COMPONENT

  A simple Svelte component for use in tests.

  NOTE:
    In the code, object properties starting with _ are safe to mangle; the name
    can be shortened at build time for smaller JavaScript bundles.
-->

<script>
  export let name = 'Elon Musk';

  $: reversed = name.split('').reverse().join('');

  // refs
  let target;
  let nameReversed;
</script>

<style type=\\"text/postcss\\">
  @media screen and (min-width: 30em) {
  :global(body) {
      margin: 0
  }
    }

  .text {
    color: rgb(255, 0, 0);
  }

  /* red */

  .text:focus {
      background: rgb(0, 0, 0); /* black */
    }

  .wrapper > .text {
      margin-top: 1rem;
    }
</style>

<svelte:options tag=\\"test-component\\" immutable={true} />

<div class=\\"wrapper\\">
  <div class=\\"text\\" bind:this=\\"{target}\\">
    test {name}
  </div>

  <div class=\\"text\\" bind:this=\\"{nameReversed}\\">
    test <span>{reversed}</span>
  </div>
</div>
"
`;

exports[`Svelte style preprocessor processes a simple component 1`] = `
"
<div id=\\"target\\">test</div>

<style type=\\"text/postcss\\">
  #target {
    color: red;
  }

    #target:focus {
      background: #000;
    }
</style>
"
`;
